modules/shared/src/main/scala/org/tessellation/schema/swap.scala

```diff
@@ -0,0 +1,72 @@
+package org.tessellation.schema
+
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.transaction.{TransactionAmount, TransactionSalt}
+import org.tessellation.security.hash.Hash
+
+import derevo.cats.{order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import eu.timepit.refined.types.numeric.PosLong
+import io.estatico.newtype.macros.newtype
+
+object swap {
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class AllowSpendFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class CurrencyId(value: Option[Address])
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class AllowSpendReference(value: Hash) // exact type TBD (e.g. lastRef?)
+
+  @derive(decoder, encoder, order, show)
+  case class AllowSpend(
+    source: Address,
+    destination: Address,
+    currency: CurrencyId,
+    amount: TransactionAmount,
+    fee: AllowSpendFee,
+    parent: AllowSpendReference,
+    salt: TransactionSalt,
+    lastValidOrdinal: SnapshotOrdinal,
+    approverAddresses: List[Address]
+  )
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapAMMDataUpdateFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  case class PriceRange(min: TransactionAmount, max: TransactionAmount)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class PoolId(value: Address) // exact type TBD
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapReference(value: Address) // exact type TBD
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SpendTransactionFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  case class SpendTransaction(
+    fee: SpendTransactionFee,
```

Comment:
IMO there should be no differentiation for transaction type. We already had `DAGTransaction` and `CurrencyTransaction` that we decided to merge under just `Transaction` and use composition over inheritance to extend the functionalities. The same should be applied to swaps

Reply:
There are some pretty significant differences between SpendTransaction and Transaction. The most important one imo is that SpendTransaction is Signed by the approver(s) of an AllowSpend, not the address transferring the tokens. It also has a required reference to an allowSpend which is needed for validation of the `amount` field. Do you think those aspects would still be best supported as Transaction w/some metadata?

Reply:
> The most important one imo is that SpendTransaction is Signed by the approver(s) of an AllowSpend, not the address transferring the tokens

The `Transaction` itself is just a data class. We have a separate `Signed[A]` wrapper and the signing process can be defined differently for various purposes. What you wrote is just the validation function -> we would require this or that signature. Anyway, it's disjoint

> It also has a required reference to an allowSpend which is needed for validation of the amount field.

This can be a metadata part indeed.

---

If the SpendTransaction should be handled and managed differently than a normal Transaction, then it's okay to make it a totally separate thing. If the only difference is some additional validation based on the metadata, I see no point for making it separate.

Reply:
Seems reasonable. I'll let others comment on it but I don't have an issue with that structure.

Reply:
@ryle-goehausen ?

Reply:
@marcinwadon @AlexBrandes @ryle-goehausen 

Are there any new/updated thoughts on this?

I see Marcin's concern about previously merging `DAGTransaction` and `CurrencyTransaction`. Isn't `SpendTransaction` a new/different thing? Isn't the flow going to be entirely different?

It feels to me that adding metadata to `Transaction` instead of creating a new class may introduce problems now or in the future with trying to assign meaning to existing `Transaction` fields that don't really fit or don't exist for `SpendTransaction`. E.g.  how do we set `Transaction.parent` ?

Reply:
As discussed in meeting, I think it doesn't make sense to re-use the old transaction class. We can discuss creating a more generic transaction in the future, but the previous schema is locked and we want to migrate away from that. For now, we should follow updates to this by handling it as a special case and then discuss a way to handle this more generically in the future.